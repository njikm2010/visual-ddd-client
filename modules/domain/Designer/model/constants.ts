export enum DomainDesignerTabs {
  Product = 'product',
  Vision = 'vision',
  UbiquitousLanguage = 'ubiquitousLanguage',
  DomainModel = 'domainModel',
  QueryModel = 'queryModel',
  DataModel = 'dataModel',
  Mapper = 'mapper',
}

export const DomainDesignerTabsMap = {
  [DomainDesignerTabs.Product]: '需求资料',
  [DomainDesignerTabs.Vision]: '产品愿景',
  [DomainDesignerTabs.UbiquitousLanguage]: '统一语言',
  [DomainDesignerTabs.DomainModel]: '领域模型',
  [DomainDesignerTabs.QueryModel]: '查询模型',
  [DomainDesignerTabs.DataModel]: '数据模型',
  [DomainDesignerTabs.Mapper]: '对象结构映射',
};
