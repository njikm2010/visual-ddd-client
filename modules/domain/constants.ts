/**
 * YJS 字段
 */
export enum YJS_FIELD_NAME {
  // 产品资料
  PRODUCT = 'product',
  // 产品愿景
  VISION = 'vision',
  // 统一语言
  UBIQUITOUS_LANGUAGE = 'ubiquitous-language',
  // 领域模型
  DOMAIN = 'domain',
  // 查询模型
  QUERY = 'query',
  // 数据对象
  DATA_OBJECT = 'data-object',
  // 对象映射
  MAPPER = 'mapper',
}
