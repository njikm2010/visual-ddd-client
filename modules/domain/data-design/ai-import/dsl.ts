import { DataObjectReferenceCardinality, DataObjectTypeName } from '../dsl/dsl';

export namespace AIImport {
  export interface DataObjectReferenceType {
    type: DataObjectTypeName.Reference;
    target: string;
    property: string;
    cardinality: DataObjectReferenceCardinality;
  }

  export interface DataObjectNormalType {
    type: Exclude<DataObjectTypeName, DataObjectTypeName.Reference>;
  }

  export interface DataObjectProperty {
    name: string;
    title: string;
    primaryKey?: boolean;
    type: DataObjectNormalType | DataObjectReferenceType;
  }

  export interface DataObject {
    name: string;
    title: string;
    properties: DataObjectProperty[];
  }
}
