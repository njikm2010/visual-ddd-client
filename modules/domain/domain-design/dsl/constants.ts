/**
 * Java 中 Void, 用于注解空 (null) 的值
 */
export const VoidClass = 'Void';

/**
 * Java 中 void 关键字
 */
export const Void = 'void';

export const UntitledInCamelCase = 'untitled';
export const UntitledInUpperCamelCase = 'Untitled';
export const UntitledInTitle = '未命名';

export const AccessModifier = {
  public: '',
  private: '-',
  protected: '#',
};

export const AccessList = [
  { label: 'public(公开)', value: 'public' },
  { label: 'private(私有)', value: 'private' },
  { label: 'protected(保护)', value: 'protected' },
];
