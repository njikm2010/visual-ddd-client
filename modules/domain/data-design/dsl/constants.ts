import { DataObjectIndexMethod, DataObjectIndexType, DataObjectReferenceCardinality, DataObjectTypeName } from './dsl';

export enum DataObjectName {
  DataObject = 'dataObject',
}

export const DataObjectReadableName: Record<DataObjectName, string> = {
  [DataObjectName.DataObject]: '数据对象',
};

export const DataObjectTypeList: { name: DataObjectTypeName; label: string }[] = [
  { name: DataObjectTypeName.Boolean, label: '布尔型' },
  { name: DataObjectTypeName.Integer, label: '整型' },
  { name: DataObjectTypeName.Decimal, label: '十进制' },
  { name: DataObjectTypeName.Long, label: '长整型' },
  { name: DataObjectTypeName.Double, label: '双精度' },
  { name: DataObjectTypeName.Float, label: '浮点型' },
  { name: DataObjectTypeName.Date, label: '日期型' },
  { name: DataObjectTypeName.DateTime, label: '日期时间型' },
  { name: DataObjectTypeName.Timestamp, label: '时间戳型' },
  { name: DataObjectTypeName.String, label: '字符串' },
  { name: DataObjectTypeName.Text, label: '文本型' },
  { name: DataObjectTypeName.LongText, label: '长文本型' },
  { name: DataObjectTypeName.JSON, label: 'JSON型' },
  { name: DataObjectTypeName.Reference, label: '引用型' },
];

export const DATE_FORMAT = 'YYYY-MM-DD';
export const DATE_TIME_FORMAT = 'YYYY-MM-DD HH:mm:ss';

export const DataObjectIndexTypeName: Record<DataObjectIndexType, string> = {
  [DataObjectIndexType.Normal]: '普通索引',
  [DataObjectIndexType.Unique]: '唯一索引',
  [DataObjectIndexType.FullText]: '全文索引',
};

export const DataObjectIndexListList: { name: DataObjectIndexType; label: string }[] = [
  { name: DataObjectIndexType.Normal, label: DataObjectIndexTypeName[DataObjectIndexType.Normal] },
  { name: DataObjectIndexType.Unique, label: DataObjectIndexTypeName[DataObjectIndexType.Unique] },
  { name: DataObjectIndexType.FullText, label: DataObjectIndexTypeName[DataObjectIndexType.FullText] },
];

export const DataObjectIndexMethodList: { name: DataObjectIndexMethod; label: string }[] = [
  { name: DataObjectIndexMethod.BTREE, label: 'BTREE' },
  { name: DataObjectIndexMethod.HASH, label: 'HASH' },
];

/**
 * 支持主键的字段类型
 * @param type
 * @returns
 */
export function objectTypeThatSupportPrimaryKey(type: DataObjectTypeName) {
  return type !== DataObjectTypeName.JSON;
}

/**
 * 支持默认值的字段类型
 * @param type
 * @returns
 */
export function objectTypeThatSupportDefaultValue(type: DataObjectTypeName) {
  return (
    type === DataObjectTypeName.Boolean ||
    type === DataObjectTypeName.Integer ||
    type === DataObjectTypeName.Decimal ||
    type === DataObjectTypeName.Long ||
    type === DataObjectTypeName.Double ||
    type === DataObjectTypeName.Float ||
    type === DataObjectTypeName.Date ||
    type === DataObjectTypeName.DateTime ||
    type === DataObjectTypeName.Timestamp ||
    type === DataObjectTypeName.String
  );
}

/**
 * 反方向基数
 */
export const DataObjectReferenceCardinalityReversed: Record<
  DataObjectReferenceCardinality,
  DataObjectReferenceCardinality
> = {
  [DataObjectReferenceCardinality.ManyToMany]: DataObjectReferenceCardinality.ManyToMany,
  [DataObjectReferenceCardinality.OneToOne]: DataObjectReferenceCardinality.OneToOne,
  [DataObjectReferenceCardinality.OneToMany]: DataObjectReferenceCardinality.ManyToOne,
  [DataObjectReferenceCardinality.ManyToOne]: DataObjectReferenceCardinality.OneToMany,
};

/**
 * 可被引用的字段类型
 * @param type
 * @returns
 */
export function objectTypeThatReferable(type: DataObjectTypeName) {
  return type !== DataObjectTypeName.Reference && type !== DataObjectTypeName.JSON;
}

/**
 * 支持递增的字段类型
 * @param type
 * @returns
 */
export function objectTypeThatSupportAutoIncrement(type: DataObjectTypeName) {
  return (
    type === DataObjectTypeName.Integer ||
    type === DataObjectTypeName.Long ||
    type === DataObjectTypeName.Double ||
    type === DataObjectTypeName.Float
  );
}

export function objectTypeThatSupportLength(type: DataObjectTypeName) {
  return type === DataObjectTypeName.String;
}
