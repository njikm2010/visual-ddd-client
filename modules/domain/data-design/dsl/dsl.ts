import { NameDSL } from '@/modules/domain/domain-design/dsl/dsl';

/**
 * 类型
 */
export enum DataObjectTypeName {
  Boolean = 'Boolean',
  Date = 'Date',
  DateTime = 'DateTime',
  Timestamp = 'Timestamp',
  Integer = 'Integer',
  Decimal = 'Decimal',
  Long = 'Long',
  Double = 'Double',
  Float = 'Float',
  String = 'String',
  Text = 'Text',
  LongText = 'LongText',
  JSON = 'JSON',
  Reference = 'Reference',
}

/**
 * 布尔值
 */
export type DataObjectBoolean = {
  type: DataObjectTypeName.Boolean;
  defaultValue?: boolean;
};

/**
 * 日期
 */
export type DataObjectDate = {
  type: DataObjectTypeName.Date;
  defaultValue?: `${number}-${number}-${number}`;
};

/**
 * 日期时间
 */
export type DataObjectDateTime = {
  type: DataObjectTypeName.DateTime;
  defaultValue?: `${number}-${number}-${number} ${number}:${number}:${number}`;
};

/**
 * 时间戳
 */
export type DataObjectTimestamp = {
  type: DataObjectTypeName.Timestamp;
  /**
   * 从 1970-01-01 00:00:00 开始的时间戳
   */
  defaultValue?: `${number}-${number}-${number} ${number}:${number}:${number}`;
};

/**
 * 整数
 */
export type DataObjectInteger = {
  type: DataObjectTypeName.Integer;
  defaultValue?: number;

  /**
   * 是否自增
   */
  autoIncrement?: boolean;
};

/**
 * 小数
 */
export type DataObjectDecimal = {
  type: DataObjectTypeName.Decimal;
  defaultValue?: number;

  /**
   * 精度, 默认 10, 最大为 65
   */
  precision?: number;

  /**
   * 小数位, 默认 0
   */
  scale?: number;
};

/**
 * 长整型
 */
export type DataObjectLong = {
  type: DataObjectTypeName.Long;
  defaultValue?: number;

  /**
   * 是否自增
   */
  autoIncrement?: boolean;
};

/**
 * Double
 */
export type DataObjectDouble = {
  type: DataObjectTypeName.Double;
  defaultValue?: number;

  /**
   * 是否自增
   */
  autoIncrement?: boolean;
};

/**
 * Float
 */
export type DataObjectFloat = {
  type: DataObjectTypeName.Float;
  defaultValue?: number;

  /**
   * 是否自增
   */
  autoIncrement?: boolean;
};

/**
 * String
 */
export type DataObjectString = {
  type: DataObjectTypeName.String;
  defaultValue?: string;

  /**
   * 对应 varchar
   * 0 to 65,535.
   * 默认 255
   */
  length?: number;
};

/**
 * Text
 */
export type DataObjectText = {
  type: DataObjectTypeName.Text;
};

/**
 * 长文本
 */
export type DataObjectLongText = {
  type: DataObjectTypeName.LongText;
};

export enum DataObjectReferenceCardinality {
  OneToOne = 'OneToOne',
  OneToMany = 'OneToMany',
  ManyToOne = 'ManyToOne',
  ManyToMany = 'ManyToMany',
}

/**
 * 字段引用
 */
export type DataObjectReference = {
  type: DataObjectTypeName.Reference;

  /**
   * 目标对象 ID
   */
  target?: string;

  /**
   * 目标对象字段 ID
   */
  targetProperty?: string;

  /**
   * 基数关系
   */
  cardinality?: DataObjectReferenceCardinality;
};

export type DataObjectJSON = {
  type: DataObjectTypeName.JSON;
};

/**
 * 所有支持的数据类型
 */
export type DataObjectTypeDSL =
  | DataObjectBoolean
  | DataObjectDate
  | DataObjectDateTime
  | DataObjectTimestamp
  | DataObjectInteger
  | DataObjectDecimal
  | DataObjectLong
  | DataObjectDouble
  | DataObjectFloat
  | DataObjectString
  | DataObjectText
  | DataObjectLongText
  | DataObjectJSON
  | DataObjectReference;

/**
 * 数据对象字段
 */
export interface DataObjectPropertyDSL<T extends DataObjectTypeDSL = DataObjectTypeDSL> extends NameDSL {
  /**
   * 表字段名，默认为 name 的 kebab_case 模式
   */
  propertyName?: string;

  /**
   * 字段类型
   */
  type: T;

  /**
   * 是否非空
   */
  notNull?: boolean;

  /**
   * 是否为主键
   */
  primaryKey?: boolean;
}

export enum DataObjectIndexType {
  Normal = 'Normal',
  Unique = 'Unique',
  // PrimaryKey = 'Primary',
  FullText = 'FullText',
}

export enum DataObjectIndexMethod {
  BTREE = 'BTREE',
  HASH = 'HASH',
}

/**
 * 索引描述
 */
export interface DataObjectIndexDSL extends NameDSL {
  /**
   * 索引类型
   */
  type: DataObjectIndexType;

  /**
   * 索引字段
   */
  properties: string[];

  /**
   * 索引方法
   */
  method: DataObjectIndexMethod;
}

/**
 * 数据对象
 */
export interface DataObjectDSL extends NameDSL {
  /**
   * 表名称，默认为 name 的 kebab_case 模式
   */
  tableName?: string;

  /**
   * 字段
   */
  properties: DataObjectPropertyDSL[];

  /**
   * 索引
   */
  indexes: DataObjectIndexDSL[];
}
