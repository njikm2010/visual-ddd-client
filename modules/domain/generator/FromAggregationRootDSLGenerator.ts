import { assert } from '@/lib/utils';
import { EntityDSL, QueryDSL } from '../domain-design/dsl/dsl';
import { createIDDSL, createQuery } from '../domain-design/dsl/factory';
import { createDataObjectDSL, createDataObjectPropertyDSL } from '../data-design/dsl/factory';
import { DataObjectDSL } from '../data-design/dsl/dsl';
import { createFieldMapperDSL, createObjectMapperDSL } from '../mapper-design/dsl/factory';
import { MapperObjectDSL, ObjectReferenceSource } from '../mapper-design/dsl/dsl';

import { BaseGenerator } from './BaseGenerator';
import { BaseGeneratorState } from './BaseGeneratorState';
import { DTOGeneratorState, IDomainObjectStore } from './types';
import { transformTypeDSLToDataObjectTypeDSL } from './transform';
import { FromEntityDSLGenerator } from './FromEntityDSLGenerator';

/**
 * 聚合根自动生成
 */
export class FromAggregationRootDSLGenerator extends BaseGenerator {
  private root: EntityDSL;
  protected domainObjectStore: IDomainObjectStore;
  protected dtoGeneratorState: DTOGeneratorState = new BaseGeneratorState();

  constructor(inject: { entity: EntityDSL; domainObjectStore: IDomainObjectStore }) {
    super();

    const { entity, domainObjectStore } = inject;
    assert(entity.isAggregationRoot, '必须以聚合根为入口进行转换');

    this.root = entity;
    this.domainObjectStore = domainObjectStore;
  }

  generate() {
    const query = this.toQuery();
    const dtos = this.dtoGeneratorState.list;
    const dataObject = this.toDataObject();
    const mapper = this.toMapper(dataObject);

    return {
      query,
      dtos,
      dataObject,
      mapper,
    };
  }

  /**
   * 转换为查询对象
   * @param entity
   */
  protected toQuery(): { detailQuery: QueryDSL; pageQuery: QueryDSL } {
    const root = this.root;

    const dtoTypeDSL = new FromEntityDSLGenerator({
      entity: root,
      dtoGeneratorState: this.dtoGeneratorState,
      queryTypeDSLTransformer: this,
    }).toDTOTypeDSL();

    const entityIdProperty = root.properties.find(i => i.uuid === root.id)!;
    const id = {
      ...entityIdProperty,
      ...createIDDSL(),
      type: entityIdProperty.type && this.transformQueryTypeDSL(entityIdProperty.type),
    };

    // 详情查询
    const detailQuery = createQuery();
    detailQuery.name = `${root.name}Query`;
    detailQuery.title = `${root.title}详情查询对象`;
    detailQuery.description = `查询${root.title}详情`;
    detailQuery.properties = [id];
    detailQuery.result = dtoTypeDSL;

    const pageQuery = createQuery();
    pageQuery.name = `${root.name}PageQuery`;
    pageQuery.title = `${root.title}分页查询对象`;
    pageQuery.description = `分页查询${root.title}`;
    pageQuery.properties = [];
    pageQuery.result = dtoTypeDSL;

    return { detailQuery, pageQuery };
  }

  /**
   * 转换为数据对象
   * @param entity
   */
  protected toDataObject(): DataObjectDSL {
    const root = this.root;

    const dataObject = createDataObjectDSL();
    dataObject.name = `${root.name}DO`;
    dataObject.title = root.title;
    dataObject.description = root.description;
    dataObject.properties = [];

    // 属性转换
    for (const property of root.properties) {
      const type =
        property.type &&
        transformTypeDSLToDataObjectTypeDSL(
          property.type,
          this.domainObjectStore.getObjectById.bind(this.domainObjectStore)
        );

      if (type == null) {
        continue;
      }

      const dataObjectProperty = createDataObjectPropertyDSL();
      dataObjectProperty.name = property.name;
      dataObjectProperty.title = property.title;
      dataObjectProperty.description = property.description;
      dataObjectProperty.type = type;

      if (root.id === property.uuid) {
        // 实体 标识符, 作为主键
        dataObjectProperty.primaryKey = true;
        dataObjectProperty.notNull = true;
      }

      dataObject.properties.push(dataObjectProperty);
    }

    return dataObject;
  }

  /**
   * 转换为对象映射
   * 需要在数据对象转换完成后进行
   */
  protected toMapper(target: DataObjectDSL): MapperObjectDSL {
    const root = this.root;

    const mapper = createObjectMapperDSL();
    mapper.name = `${root.name}To${target.name}`;
    mapper.title = `${root.title}数据对象映射`;
    mapper.description = `用于将聚合根(${root.title})映射为${target.title}(数据对象)`;

    mapper.source = {
      source: ObjectReferenceSource.Domain,
      referenceId: root.uuid,
      name: root.name,
    };

    mapper.target = {
      source: ObjectReferenceSource.Data,
      referenceId: target.uuid,
      name: target.name,
    };

    // 自动字段映射
    for (const property of root.properties) {
      const targetProperty = target.properties.find(p => p.name === property.name);
      if (targetProperty == null) {
        continue;
      }

      const fieldMapper = createFieldMapperDSL();
      fieldMapper.source = property.uuid;
      fieldMapper.target = targetProperty.uuid;

      mapper.mappers.push(fieldMapper);
    }

    return mapper;
  }
}
