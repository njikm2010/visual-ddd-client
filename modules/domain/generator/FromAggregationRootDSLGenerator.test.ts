import * as mockedUuid from 'uuid';
import { DomainObjectName } from '../domain-design/dsl/constants';
import {
  createBaseType,
  createContainerType,
  createEntity,
  createEnum,
  createProperty,
  createReferenceType,
  createValueObject,
} from '../domain-design/dsl/factory';
import { FromAggregationRootDSLGenerator } from './FromAggregationRootDSLGenerator';

jest.mock('uuid', () => {
  let count = 0;
  return {
    __esModule: true,
    reset() {
      count = 0;
    },
    v4() {
      return `uuid-${count++}`;
    },
  };
});

const mockedUuidModule = mockedUuid as unknown as jest.Mocked<{ reset(): void }>;

describe('FromAggregationRootDSLGenerator', () => {
  beforeEach(() => {
    mockedUuidModule.reset();
  });

  it('should throw error, if entity is not aggregation root', () => {
    const entity = createEntity();
    entity.isAggregationRoot = false;

    expect(() => {
      new FromAggregationRootDSLGenerator({
        entity,
        domainObjectStore: {
          getObjectById() {
            return undefined;
          },
        },
      });
    }).toThrowError();
  });

  it('generate', () => {
    const entity = createEntity();
    entity.name = 'MyEntity';
    entity.title = '我的实体';
    entity.description = '我的实体描述';
    entity.isAggregationRoot = true;

    entity.properties = [];

    // 基础类型
    const baseTypeProperty = createProperty();
    baseTypeProperty.type = createBaseType('String');
    baseTypeProperty.name = 'baseTypeProperty';
    baseTypeProperty.title = '基础类型属性';
    baseTypeProperty.description = '基础类型属性的描述';

    const baseType2Property = createProperty();
    baseType2Property.type = createBaseType('Long');
    baseType2Property.name = 'baseType2Property';
    baseType2Property.title = '基础类型属性2';
    baseType2Property.description = '基础类型属性的描述2';

    // 容器类型
    const containerTypeProperty = createProperty();
    containerTypeProperty.type = createContainerType('List');
    containerTypeProperty.type.params.item = createBaseType('Boolean');
    containerTypeProperty.name = 'containerTypeProperty';

    // 引用类型 enum
    const referenceTypePropertyForEnum = createProperty();
    referenceTypePropertyForEnum.type = createReferenceType(DomainObjectName.Enum, 'enumName');
    referenceTypePropertyForEnum.name = 'referenceTypePropertyForEnum';

    // 引用类型 valueObject
    const referenceTypePropertyForValueObject = createProperty();
    referenceTypePropertyForValueObject.type = createReferenceType(DomainObjectName.ValueObject, 'valueObjectName');
    referenceTypePropertyForValueObject.name = 'referenceTypePropertyForValueObject';

    // 引用类型 entity
    const referenceTypePropertyForEntity = createProperty();
    referenceTypePropertyForEntity.type = createReferenceType(DomainObjectName.Entity, 'entityName');
    referenceTypePropertyForEntity.name = 'referenceTypePropertyForEntity';

    entity.properties.push(
      baseTypeProperty,
      baseType2Property,
      containerTypeProperty,
      referenceTypePropertyForEnum,
      referenceTypePropertyForValueObject,
      referenceTypePropertyForEntity
    );

    entity.id = baseTypeProperty.uuid;

    const gen = new FromAggregationRootDSLGenerator({
      entity,
      domainObjectStore: {
        getObjectById(id) {
          switch (id) {
            case DomainObjectName.Enum: {
              const obj = createEnum();
              obj.name = 'enumName';
              obj.title = '枚举';
              obj.description = '枚举描述';

              return {
                type: DomainObjectName.Enum,
                value: obj,
              };
            }
            case DomainObjectName.ValueObject: {
              const obj = createValueObject();
              obj.name = 'valueObjectName';
              obj.title = '值对象';
              obj.description = '值对象描述';

              return {
                type: DomainObjectName.ValueObject,
                value: obj,
              };
            }
            case DomainObjectName.Entity: {
              const obj = createEntity();
              obj.name = 'entityName';
              obj.title = '实体';
              obj.description = '实体描述';

              return {
                type: DomainObjectName.Entity,
                value: obj,
              };
            }
          }
        },
      },
    });

    const result = gen.generate();

    expect(result).toMatchSnapshot();
  });
});
