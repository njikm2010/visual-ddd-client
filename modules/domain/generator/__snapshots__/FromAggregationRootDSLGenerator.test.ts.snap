// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FromAggregationRootDSLGenerator generate 1`] = `
{
  "dataObject": {
    "description": "我的实体描述",
    "indexes": [],
    "meta": [],
    "name": "MyEntityDO",
    "properties": [
      {
        "description": "基础类型属性的描述",
        "meta": [],
        "name": "baseTypeProperty",
        "notNull": true,
        "primaryKey": true,
        "propertyName": undefined,
        "title": "基础类型属性",
        "type": {
          "type": "String",
        },
        "uuid": "uuid-44",
      },
      {
        "description": "基础类型属性的描述2",
        "meta": [],
        "name": "baseType2Property",
        "notNull": false,
        "primaryKey": false,
        "propertyName": undefined,
        "title": "基础类型属性2",
        "type": {
          "type": "Long",
        },
        "uuid": "uuid-45",
      },
      {
        "description": "",
        "meta": [],
        "name": "referenceTypePropertyForEnum",
        "notNull": false,
        "primaryKey": false,
        "propertyName": undefined,
        "title": "未命名",
        "type": {
          "type": "Integer",
        },
        "uuid": "uuid-48",
      },
    ],
    "tableName": undefined,
    "title": "我的实体",
    "uuid": "uuid-42",
  },
  "dtos": [
    {
      "abstract": false,
      "classMethods": [],
      "classProperties": [],
      "description": "用于我的实体的数据传输对象",
      "extends": undefined,
      "implements": [],
      "meta": [],
      "methods": [],
      "name": "MyEntityDTO",
      "properties": [
        {
          "access": "public",
          "description": "基础类型属性的描述",
          "meta": [],
          "name": "baseTypeProperty",
          "title": "基础类型属性",
          "type": {
            "name": "String",
            "type": "base",
          },
          "uuid": "uuid-13",
        },
        {
          "access": "public",
          "description": "基础类型属性的描述2",
          "meta": [],
          "name": "baseType2Property",
          "title": "基础类型属性2",
          "type": {
            "name": "Long",
            "type": "base",
          },
          "uuid": "uuid-14",
        },
        {
          "access": "public",
          "description": "",
          "meta": [],
          "name": "containerTypeProperty",
          "title": "未命名",
          "type": {
            "name": "List",
            "params": {
              "item": {
                "name": "Boolean",
                "type": "base",
              },
            },
            "type": "container",
          },
          "uuid": "uuid-15",
        },
        {
          "access": "public",
          "description": "",
          "meta": [],
          "name": "referenceTypePropertyForEnum",
          "title": "未命名",
          "type": {
            "name": "Integer",
            "type": "base",
          },
          "uuid": "uuid-16",
        },
        {
          "access": "public",
          "description": "",
          "meta": [],
          "name": "referenceTypePropertyForValueObject",
          "title": "未命名",
          "type": {
            "name": "valueObjectNameDTO",
            "referenceId": "uuid-23",
            "type": "reference",
          },
          "uuid": "uuid-19",
        },
        {
          "access": "public",
          "description": "",
          "meta": [],
          "name": "referenceTypePropertyForEntity",
          "title": "未命名",
          "type": {
            "name": "entityNameDTO",
            "referenceId": "uuid-32",
            "type": "reference",
          },
          "uuid": "uuid-27",
        },
      ],
      "title": "我的实体数据传输对象",
      "uuid": "uuid-10",
    },
    {
      "abstract": false,
      "classMethods": [],
      "classProperties": [],
      "description": "用于值对象的数据传输对象",
      "extends": undefined,
      "implements": [],
      "meta": [],
      "methods": [],
      "name": "valueObjectNameDTO",
      "properties": [
        {
          "access": "public",
          "description": "",
          "meta": [],
          "name": "untitled",
          "title": "未命名",
          "type": {
            "name": "String",
            "type": "base",
          },
          "uuid": "uuid-26",
        },
      ],
      "title": "值对象数据传输对象",
      "uuid": "uuid-23",
    },
    {
      "abstract": false,
      "classMethods": [],
      "classProperties": [],
      "description": "用于实体的数据传输对象",
      "extends": undefined,
      "implements": [],
      "meta": [],
      "methods": [],
      "name": "entityNameDTO",
      "properties": [
        {
          "access": "public",
          "description": "",
          "meta": [],
          "name": "id",
          "title": "实体唯一标识符",
          "type": {
            "name": "Long",
            "type": "base",
          },
          "uuid": "uuid-35",
        },
        {
          "access": "public",
          "description": "",
          "meta": [],
          "name": "untitled",
          "title": "未命名",
          "type": {
            "name": "String",
            "type": "base",
          },
          "uuid": "uuid-36",
        },
      ],
      "title": "实体数据传输对象",
      "uuid": "uuid-32",
    },
  ],
  "mapper": {
    "description": "用于将聚合根(我的实体)映射为我的实体(数据对象)",
    "mappers": [
      {
        "source": "uuid-4",
        "target": "uuid-44",
        "uuid": "uuid-57",
      },
      {
        "source": "uuid-5",
        "target": "uuid-45",
        "uuid": "uuid-58",
      },
      {
        "source": "uuid-7",
        "target": "uuid-48",
        "uuid": "uuid-59",
      },
    ],
    "meta": [],
    "name": "MyEntityToMyEntityDO",
    "source": {
      "name": "MyEntity",
      "referenceId": "uuid-0",
      "source": "domain",
    },
    "target": {
      "name": "MyEntityDO",
      "referenceId": "uuid-42",
      "source": "data",
    },
    "title": "我的实体数据对象映射",
    "uuid": "uuid-56",
  },
  "query": {
    "detailQuery": {
      "description": "查询我的实体详情",
      "meta": [],
      "name": "MyEntityQuery",
      "pagination": false,
      "properties": [
        {
          "access": "public",
          "description": "基础类型属性的描述",
          "meta": [],
          "name": "baseTypeProperty",
          "title": "基础类型属性",
          "type": {
            "name": "String",
            "type": "base",
          },
          "uuid": "uuid-37",
        },
      ],
      "result": {
        "name": "MyEntityDTO",
        "referenceId": "uuid-10",
        "type": "reference",
      },
      "source": {
        "event": {
          "enabled": false,
          "value": "",
        },
        "http": {
          "enabled": true,
        },
        "rpc": {
          "enabled": true,
        },
        "schedule": {
          "enabled": false,
          "value": "",
        },
      },
      "title": "我的实体详情查询对象",
      "uuid": "uuid-38",
    },
    "pageQuery": {
      "description": "分页查询我的实体",
      "meta": [],
      "name": "MyEntityPageQuery",
      "pagination": true,
      "properties": [],
      "result": {
        "name": "MyEntityDTO",
        "referenceId": "uuid-10",
        "type": "reference",
      },
      "source": {
        "event": {
          "enabled": false,
          "value": "",
        },
        "http": {
          "enabled": true,
        },
        "rpc": {
          "enabled": true,
        },
        "schedule": {
          "enabled": false,
          "value": "",
        },
      },
      "title": "我的实体分页查询对象",
      "uuid": "uuid-40",
    },
  },
}
`;
